### PR TITLE
[PEAA-169] Add tag element

### PR DESCRIPTION
**Pull request by tradeshift-larsbojensen**
I didn't add storybook nor a happo test. I want to discuss and possibly merge the component first

### CODE REVIEW

File: packages/components/tag/src/utils/constants.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+export const labels = {
+	SINGLE_LABEL: ['Label 1'],
+	MULTIPLE_LABELS: ['Label 1', 'Label 2']
+};
+
+export const values = {
+	SINGLE_VALUE: ['value 1'],
+	MULTIPLE_VALUES: ['value 1', 'value 2']
+};
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/utils/constants.js:18]: I think we should not use the utils for the stories-only usage and everything related to the stories should live inside that directory. In this case I would put it directly to the happo file.

File: packages/components/tag/src/tag.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			/** Text direction: 'rtl' or 'ltr' */
+			dir: { type: String, reflect: true },
+			/** Type: success, warning, warning-lite, danger, blue, blue-lite */
+			type: { type: String, reflect: true },
+			/** Can the item be deleted? */
+			deletable: { type: Boolean, reflect: true },
+			/** Should the tag look like clicking it triggers an action? */
+			clickable: { type: Boolean, reflect: true },
+			/** Is the item locked? */
+			locked: { type: Boolean, reflect: true },
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.js:21]: Just a general suggestion, I think we should also try to mention what would be reflected in the UI or behaviour by changing these kinda values.

File: packages/components/tag/src/tag.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		/** constraint event; do not trigger events up the DOM tree */
+		event.stopPropagation();
+		if (!this.busy) {
+			this.dispatchCustomEvent('delete-tag', event);
+		}
+	}
+
+	handleClickEvent(event) {
+		if (!this.busy) {
+			this.dispatchCustomEvent('click-tag', event);
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.js:45]: Please add a comment, so our doc generator would pick it up for our readme file, [see here](https://github.com/Tradeshift/elements/blob/master/packages/components/search/src/search.js#L140)

File: packages/components/tag/src/tag.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+	get direction() {
+		return this.dir || this.bodyDir;
+	}
+
+	handleDeleteEvent(event) {
+		/** constraint event; do not trigger events up the DOM tree */
+		event.stopPropagation();
+		if (!this.busy) {
+			this.dispatchCustomEvent('delete-tag', event);
```
[the remark this answers is not in the document]
Comment (esmailpour-hosein) [reply packages/components/tag/src/tag.js:39]: I think we can also send some of tag information like key and value, so developers can have one listener and distinguish items without looking into target detail, if needed.

File: packages/components/tag/src/tag.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		}
+
+		if (this.locked) {
+			return html`
+				<span class="icon">
+					<ts-icon icon="lock" size="large"></ts-icon>
+				</span>
+			`;
+		}
+
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.js:74]: by swapping their places, we can get rid of the `!this.locked` condition on the first `if`

File: packages/components/tag/src/tag.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		return html`<div
+			dir=${this.direction}
+			class=${this.classes}
+			@click=${this.handleClickEvent}
+			@keyup=${this.handleClickEvent}
+		>
+			${this.labels &&
+			html`<span class="label">${this.labels.join(', ')}${this._hasBothLabelsAndValues() ? `:` : null}</span>`}
+			${this._hasBothLabelsAndValues() ? html`<span>&nbsp;</span>` : null}
+			${this.values && html`<span class="value">${this.values.join(', ')}</span>`} ${this.icons}
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.js:88]: We should refactor this, it's not easy to read and follow. 
Maybe we can have multiple getters like `icons` to separate these from each other and simpler render block

Comment (esmailpour-hosein) [reply packages/components/tag/src/tag.js:88]: we can do it in another PR, not a blocker

Comment (tradeshift-larsbojensen) [reply packages/components/tag/src/tag.js:88]: Yeah, I was thinking the same, but decided they were not too complicated to read, also considering the simplicity of the entire component markup. I agree it's a nice idea, though.

File: packages/components/tag/src/tag.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tradeshift-larsbojensen) [reply packages/components/tag/src/tag.js:39]: Checking keys and/or values is looking into detail, just not of the target.detail. I suggest developers add id's to ts-tags in order to identify them. But let's discuss this, maybe you are right.

File: packages/components/tag/src/utils/constants.js (hunk 1 above)
Comment (tradeshift-larsbojensen) [reply packages/components/tag/src/utils/constants.js:18]: I agree, I just thought it was a decided standard. I also think it's over the top in a case like this.

File: packages/components/tag/src/tag.css [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+/* Container base styles ***************************/
+
+:host .container {
+	display: inline-flex;
+	align-items: center;
+	margin: 0;
+	max-width: 100%;
+	min-height: 30px;
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.css:15]: please use --ts-unit-plus

File: packages/components/tag/src/tag.css [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+:host(:not([type])[clickable]) .container {
+	color: var(--ts-color-blue);
+}
+
+/* Icons ........................................................*/
+
+:host([locked]),
+:host([deletable]) {
+	& .container[dir='ltr'] {
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.css:179]: nit, Generally we have default state, which would be ltr and add specific dir='rtl' overrides.

File: packages/components/tag/src/tag.css [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+		padding-left: 0;
+	}
+}
+
+.icon {
+	display: inline-flex;
+	align-items: inherit;
+	justify-content: center;
+	width: 30px;
+	height: 30px;
```
Comment (esmailpour-hosein) [inline packages/components/tag/src/tag.css:195]: please use --ts-unit-plus

**Review by esmailpour-hosein — APPROVED** [review]
Great job! 🎉 
Nothing big, just some suggestions and minor improvements.

File: packages/components/tag/src/tag.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (esmailpour-hosein) [reply packages/components/tag/src/tag.js:39]: Let's keep it as it is for now.

**Pull request merged**